{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# C.S.I. \n",
    "### Computational Sequence Investigator\n",
    "This is a thing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO, AlignIO\n",
    "from Bio.SubsMat import MatrixInfo\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Align.Applications import ClustalwCommandline\n",
    "from Bio.Phylo.Applications import PhymlCommandline\n",
    "\n",
    "import dendropy\n",
    "from dendropy.calculate import treemeasure\n",
    "import matplotlib as mpl \n",
    "mpl.use('Agg')\n",
    "import matplotlib.pyplot as plt \n",
    "import numpy as np\n",
    "\n",
    "from random import sample\n",
    "import sys\n",
    "import os\n",
    "import glob\n",
    "import math\n",
    "import re "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###a couple things that I need to refer to####\n",
    "##MW for AAs\n",
    "my_aa = {\n",
    "\"A\": [71.0788, 0],\n",
    "\"R\": [156.1875, 0],\n",
    "\"N\": [114.1038, 0],\n",
    "\"D\": [115.0886, 0],\n",
    "\"C\": [103.1388, 0],\n",
    "\"E\": [129.1155, 0],\n",
    "\"Q\": [128.1307, 0],\n",
    "\"G\": [57.0519, 0],\n",
    "\"H\": [137.1411, 0],\n",
    "\"I\": [113.1594, 0],\n",
    "\"L\": [113.1594, 0],\n",
    "\"K\": [128.1741, 0],\n",
    "\"M\": [131.1926, 0],\n",
    "\"F\": [147.1766, 0],\n",
    "\"P\": [97.1167, 0],\n",
    "\"S\": [87.0782, 0],\n",
    "\"T\": [101.1051, 0],\n",
    "\"W\": [186.2132, 0],\n",
    "\"Y\": [163.176, 0],\n",
    "\"V\": [99.1326, 0],\n",
    "}\n",
    "##pkvalues\n",
    "my_pkvals = {\n",
    "\"Nterm\": 9.094,\n",
    "\"Cterm\": 2.869,\n",
    "\"C\": 7.555,\n",
    "\"D\": 3.872,\n",
    "\"E\": 4.412,\n",
    "\"H\": 5.637,\n",
    "\"K\": 9.052,\n",
    "\"R\": 11.84,\n",
    "\"Y\": 10.85,\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#calculate MW\n",
    "def mw(seq):\n",
    "    i = 0\n",
    "    total_MW = 0\n",
    "    #my_weights = my_aa\n",
    "    for i in range(0, len(seq)):\n",
    "        new_MW = my_aa[str(seq[i])][0]\n",
    "        total_MW = round(total_MW + new_MW,4)\n",
    "    return total_MW\n",
    "\n",
    "#calculate charge based on pH\n",
    "def charge(seq, pH):\n",
    "    i = 0\n",
    "    total_charge = 0\n",
    "    for i in range (0, len(seq)):\n",
    "        if seq[i] == \"K\" or seq[i] == \"R\" or seq[i] == \"H\":\n",
    "            partial_charge = (1/(1 + pow(10,(pH-my_pkvals[str(seq[i])]))))\n",
    "            total_charge = total_charge + partial_charge\n",
    "        elif seq[i] == \"D\" or seq[i] == \"E\" or seq[i] == \"C\" or seq[i] == \"Y\":\n",
    "            partial_charge = (-1/(1 + pow(10,(my_pkvals[str(seq[i])]-pH))))\n",
    "            total_charge = total_charge + partial_charge\n",
    "    n_term_charge = (1/(1 + pow(10,(pH-my_pkvals[\"Nterm\"]))))\n",
    "    c_term_charge = (-1/(1 + pow(10,(my_pkvals[\"Cterm\"]-pH))))\n",
    "    total_charge = n_term_charge + total_charge + c_term_charge\n",
    "    return total_charge\n",
    "\n",
    "#calculate isoelectric point\n",
    "def pI(seq):\n",
    "    test_pH = 7\n",
    "    diff = 4\n",
    "    test_charge = charge(seq, test_pH)\n",
    "    while round(test_charge, 4) != 0:\n",
    "        if test_charge > 0:\n",
    "            test_pH = test_pH + diff\n",
    "            test_charge = charge(seq, test_pH)\n",
    "            diff = float(diff/2)\n",
    "        else:\n",
    "            test_pH = test_pH - diff\n",
    "            test_charge = charge(seq, test_pH)\n",
    "            diff = float(diff/2)\n",
    "    pI_val = round(test_pH, 4)\n",
    "    return pI_val\n",
    "\n",
    "#output MW, pI and charge at 7 as a list\n",
    "def info(seq):\n",
    "    molw = mw(seq)\n",
    "    pot = pI(seq)\n",
    "    ch = round(charge(seq, 7),4)\n",
    "    results = [molw, pot, ch]\n",
    "    return results\n",
    "    \n",
    "    \n",
    "##functions to sort anything based on re match of numbers in it\n",
    "\n",
    "def return_number(text):\n",
    "    return int(text) if text.isdigit() else text\n",
    "        \n",
    "def natural_keys(text):\n",
    "    return [return_number(c) for c in re.split(r'(\\d+)', text) ]\n",
    "            \n",
    "\n",
    "#find ORF, return translation\n",
    "def translate(seq):\n",
    "\tstartcodon = re.compile(\"ATG\")\n",
    "\tstopcodon = re.compile(\"TGA|TAA|TAG\")\n",
    "\tORF = []\n",
    "\tfor match_start in startcodon.finditer(seq):\n",
    "\t\tremaining = seq[match_start.start():]\n",
    "\t\tif stopcodon.search(remaining):\n",
    "\t\t\tfor match_stop in stopcodon.finditer(remaining):\n",
    "\t\t\t\tframe = remaining[0:match_stop.end()]\n",
    "\t\t\t\tif len(frame) % 3 == 0:\n",
    "\t\t\t\t\tORF.append(frame)\n",
    "\t\t\t\t\tbreak\n",
    "\tsorted_ORF = sorted(ORF, key = len)\n",
    "\tfinal_ORF = Seq(sorted_ORF[len(sorted_ORF)-1])\n",
    "\tprotein = str(final_ORF.translate())\n",
    "\treturn protein \n",
    "\n",
    "#use dendropy to calculate pairwise distances, output a dictionary with all distances \n",
    "def distances(input):\n",
    "    the_dict = {}\n",
    "    tree = dendropy.Tree.get_from_path(input, \"newick\", preserve_underscores=True)\n",
    "    pdm = treemeasure.PatristicDistanceMatrix(tree)\n",
    "    for i, t1 in enumerate(tree.taxon_namespace):\n",
    "        tax1 = str(t1)[1:len(str(t1))-1]\n",
    "        the_dict[str(tax1)] = {}\n",
    "        for t2 in tree.taxon_namespace[i+1:]:\n",
    "            tax2 = str(t2)[1:len(str(t2))-1]\n",
    "            the_dict[str(tax1)][str(tax2)] = round(float(pdm(t1, t2)),5)\n",
    "    return the_dict\n",
    "\n",
    "#rebuild a fasta file starting from a defined dictionary, save as defined output\n",
    "def rebuild_fasta(input_dict, output_fa):\n",
    "    filename = str(output_fa) + \".fa\"\n",
    "    fileout = open(filename, \"w\")\n",
    "    for key in sorted_keys:\n",
    "        start = 0\n",
    "        end = 70\n",
    "        fileout.write (\">\"+key+\"\\n\")\n",
    "        seq = input_dict[key]\n",
    "        while end <= len(seq):\n",
    "            fileout.write(seq[start:end]+\"\\n\")\n",
    "            start = end \n",
    "            end = end + 70\n",
    "        fileout.write(seq[start:len(seq)]+\"\\n\")\n",
    "    fileout.close()\n",
    "    return filename\n",
    "\n",
    "###functions to calculate scores for protein alignment\n",
    "\n",
    "def getScore(a,b,S):\n",
    "    if (a,b)  in S:\n",
    "        return S[a,b]\n",
    "    elif (b,a) in S:\n",
    "        return S[b,a] \n",
    "    else:\n",
    "        return int(0)\n",
    "        \n",
    "def calculate_score(pep1, pep2, matrix): \n",
    "    i=0 \n",
    "    list = []\n",
    "    while i<len(pep1): \n",
    "        Score = getScore(pep1[i],pep2[i], matrix) \n",
    "        list.append(Score) \n",
    "        i = i+1 \n",
    "    return list\n",
    " \n",
    "#converts to phylip for phyml\n",
    "def convert_to_phylip(filename):\n",
    "    convert_to_phy = AlignIO.parse(filename,\"clustal\")\n",
    "    outfile = str(filename) + \".phy\"\n",
    "    AlignIO.write(convert_to_phy, open(outfile,\"w\"),\"phylip\")\n",
    "\n",
    "#extrapolates distance from distance matrices: since matrices of pw distances are not uniform, \n",
    "#this search for distance from a to b, and if not found, returns distance from b to a\n",
    "def extrapolate_distance(dictionary, tx1, tx2):\n",
    "    if tx1 in dictionary:\n",
    "        if tx2 in dictionary[tx1]:\n",
    "            #print (\"primary taxon \" + str(tx1) + \" and secondary taxon \" + str(tx2) + \" were found in the right order in dictionary \")\n",
    "            result = dictionary[taxon1][taxon2]\n",
    "        elif tx2 in dictionary:\n",
    "            if tx1 in dictionary[taxon2]:\n",
    "                #print (\"primary taxon \" + str(tx1) + \" was not found in primary keys but in secondary, so results were computed anyway\")\n",
    "                result = dictionary[taxon2][taxon1]\n",
    "    else:\n",
    "        result = 0\n",
    "    return float(result)\n",
    "\n",
    "#function to plot x and y or several x and one Y \n",
    "def my_plot(title, y, x1, x1lab, x2=None, x2lab=None, x3=None, x3lab=None, x4=None, x4lab=None):\n",
    "\n",
    "    plt.scatter(x1, y, s = 5, marker = '*', label = x1lab)\n",
    "    fit = np.polyfit(x1, y, 1)\n",
    "    p = np.poly1d(fit)\n",
    "    plt.plot(x1, p(x1), \"r--\", c = 'b')\n",
    "    \n",
    "    if x2 != None and x2lab != None:\n",
    "        plt.scatter(x2, y, s = 5, marker = 'x', label = x2lab)\n",
    "        fit = np.polyfit(x2, y, 1)\n",
    "        p = np.poly1d(fit)\n",
    "        plt.plot(x2, p(x2), \"r:\", c = 'y')\n",
    "\n",
    "    if x3 != None and x3lab != None:\n",
    "        plt.scatter(x3, y, s = 5, marker = '^', label = x3lab)\n",
    "        fit = np.polyfit(x3, y, 1)\n",
    "        p = np.poly1d(fit)\n",
    "        plt.plot(x3, p(x3), \"r-.\", c = 'g')\n",
    "    \n",
    "    if x4 != None and x4lab != None:\n",
    "        plt.scatter(x4, y, s = 5, marker = \"D\", label = x4lab)\n",
    "        fit = np.polyfit(x4, y, 1)\n",
    "        p = np.poly1d(fit)\n",
    "        plt.plot(x4, p(x4), \"r-\", c = 'r')\n",
    "    \n",
    "    plt.legend()\n",
    "    plt.savefig(title)\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Whatever"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#input statement, define input file \n",
    "InputDNA = sys.argv[1]\n",
    "    \n",
    "InputDNA_noext = InputDNA[0:len(InputDNA)-3] ##change this to regex \n",
    "DNAdict = {}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#parse DNA input, generate indexed dictionary with trimmed keys -- WORKS\n",
    "with open(InputDNA) as f:\n",
    "    for line in f:\n",
    "        if len(line) > 1:\n",
    "            if '>' in line:\n",
    "                key = line.strip()[1:11]\n",
    "                DNAdict[key] = \"\"\n",
    "            else:\n",
    "                DNAdict[key] += line.strip() \n",
    "\n",
    "#a list to sort the fasta files\n",
    "sorted_keys = []\n",
    "for key in DNAdict:\n",
    "\tsorted_keys.append(str(key))\n",
    "sorted_keys.sort(key=natural_keys)\n",
    "#print sorted_keys\n",
    "\n",
    "## translate files, store them in a dictionary. Calculate protein structure info as well\n",
    "AAdict = {}\n",
    "PROT_info_dict = {}\n",
    "for key in DNAdict:\n",
    "\tseq_str = str(DNAdict[key])\n",
    "\tPROT_info_dict[key] = info(seq_str)\n",
    "\tAAdict[key] = translate(seq_str)\n",
    "\n",
    "\n",
    "## rebuild fasta file with proteins\n",
    "output_protein_name = str(InputDNA_noext) + \"_trans\"\n",
    "AA_fasta_ref = rebuild_fasta(AAdict, output_protein_name)\n",
    "\n",
    "##run ClustalW with DNA\n",
    "\n",
    "clustalw_NUC = ClustalwCommandline(\"clustalw2\",infile=InputDNA, type='dna', quiet=True)\n",
    "os.system(str(clustalw_NUC))\n",
    "\n",
    "\n",
    "##run ClustalW with Protein\n",
    "\n",
    "clustalw_AA = ClustalwCommandline(\"clustalw2\",infile=AA_fasta_ref, type='PROTEIN', quiet=True)\n",
    "os.system(str(clustalw_AA))\n",
    "\n",
    "alnfile_DNA = str(InputDNA_noext) + \".aln\"\n",
    "alnfile_AA = str(output_protein_name) + \".aln\"\n",
    "\n",
    "###make dictionary to store DNA alignments\n",
    "alignment_dict_DNA = {}\n",
    "alignment = list(AlignIO.parse(alnfile_DNA,\"clustal\"))\n",
    "for i in range(0, len(alignment[0])):\n",
    "    id = str(alignment[0][i].id)[0:10]\n",
    "    seq = str(alignment[0][i]._seq)\n",
    "    alignment_dict_DNA[id] = seq\n",
    "\n",
    "###make dictionary to store protein alignments\n",
    "alignment_dict_AA = {}\n",
    "alignmentAA = list(AlignIO.parse(alnfile_AA,\"clustal\"))\n",
    "for i in range(0, len(alignmentAA[0])):\n",
    "    id = str(alignmentAA[0][i].id)[0:10]\n",
    "    seq = str(alignmentAA[0][i]._seq)\n",
    "    alignment_dict_AA[id] = seq\n",
    "\n",
    "#compute transitions and transversions, gaps, total score. \n",
    "#store in dictionary for each sequence pair\n",
    "\n",
    "tr_tv_dict = {}\n",
    "for key in alignment_dict_DNA:\n",
    "    taxon1 = str(key) \n",
    "    tr_tv_dict[taxon1] = {}\n",
    "    for key2 in alignment_dict_DNA:\n",
    "        taxon2 = str(key2)\n",
    "        transitions = 0 \n",
    "        transversions = 0\n",
    "        gaps = 0\n",
    "        matches = 0\n",
    "        for i in range(0, len(alignment_dict_DNA[key])):\n",
    "            pair_clean = str(alignment_dict_DNA[key][i]) + str(alignment_dict_DNA[taxon2][i])\n",
    "            if pair_clean == \"CG\" or pair_clean == \"GC\" or pair_clean == \"AT\" or pair_clean == \"TA\":\n",
    "                transversions = transversions + 1 \n",
    "            elif pair_clean == \"AC\" or pair_clean == \"CA\" or pair_clean == \"TG\" or pair_clean == \"GT\":\n",
    "                transitions = transitions + 1\n",
    "            elif pair_clean == \"A-\" or pair_clean == \"-A\"or pair_clean == \"C-\" or pair_clean == \"-C\" or pair_clean == \"G-\" or pair_clean == \"-G\" or pair_clean == \"T-\" or pair_clean == \"-T\":\n",
    "                gaps = gaps + 1 \n",
    "            elif pair_clean == \"AA\" or pair_clean == \"GG\" or pair_clean == \"CC\" or pair_clean == \"TT\":\n",
    "                matches = matches + 1\n",
    "        tot_score = (2 * matches) + ((-2) * transitions) + ((-3) * gaps) + ((-4) * transversions) \n",
    "        total = transitions + transversions\n",
    "        if total >  0:\n",
    "            transitions_of_total = float(transitions) / total\n",
    "            transversions_of_total = float(transversions) / total\n",
    "        else: \n",
    "            transitions_of_total = 0\n",
    "            transversions_of_total = 0\n",
    "        \n",
    "        tr_tv_dict[taxon1][taxon2] = [transitions, round(transitions_of_total,4), transversions, round(transversions_of_total,4), gaps, matches, tot_score]\n",
    "\n",
    "###compute BLOSUM62 and PAM60 scores based on matrix \n",
    "\n",
    "AA_score_dict = {}\n",
    "for key1 in alignment_dict_AA:\n",
    "    AA_score_dict[key1] = {}\n",
    "    for key2 in alignment_dict_AA:\n",
    "        blo = MatrixInfo.blosum62\n",
    "        pam = MatrixInfo.pam60\n",
    "        prot1 = str(alignment_dict_AA[key1]) \n",
    "        prot2 = str(alignment_dict_AA[key2]) \n",
    "        \n",
    "        blosum_list = calculate_score(prot1,prot2,blo)\n",
    "        pam_list = calculate_score(prot1,prot2,pam)\n",
    "        finalblo = sum(blosum_list)\n",
    "        finalpam = sum(pam_list)\n",
    "        AA_score_dict[key1][key2] = [finalblo, finalpam]\n",
    "\n",
    "#convert aln files to phylip\n",
    "convert_to_phylip(alnfile_DNA)\n",
    "convert_to_phylip(alnfile_AA)\n",
    "\n",
    "#run phyml on DNA\n",
    "phyfile_DNA = str(alnfile_DNA) + \".phy\"\n",
    "\n",
    "Phyml_runK = PhymlCommandline(input=phyfile_DNA, model='K80')\n",
    "os.system(str(Phyml_runK)) \n",
    "Files = glob.glob('*tree.txt') \n",
    "for file in Files: \n",
    "    os.rename(file, 'arbolK80')\n",
    "\n",
    "Phyml_runG = PhymlCommandline(input=phyfile_DNA, model='GTR')\n",
    "os.system(str(Phyml_runG))\n",
    "Files = glob.glob('*tree.txt')\n",
    "for file in Files:\n",
    "    os.rename(file,'arbolGTR')\n",
    "\n",
    "Phyml_runH = PhymlCommandline(input=phyfile_DNA, model='HKY85')\n",
    "os.system(str(Phyml_runH))\n",
    "Files = glob.glob('*tree.txt')\n",
    "for file in Files:\n",
    "    os.rename(file,'arbolHKY85')\n",
    "\n",
    "Phyml_runJC = PhymlCommandline(input=phyfile_DNA, model='JC69')\n",
    "os.system(str(Phyml_runJC))\n",
    "Files = glob.glob('*tree.txt')\n",
    "for file in Files:\n",
    "    os.rename(file,'arbolJC69')\n",
    "\n",
    "#run phyml on AA\n",
    "\n",
    "phyfile_AA = str(alnfile_AA) + \".phy\"\n",
    "\n",
    "Phyml_runBLO = PhymlCommandline(input=phyfile_AA, datatype=\"aa\", model='Blosum62')\n",
    "os.system(str(Phyml_runBLO)) \n",
    "Files = glob.glob('*tree.txt') \n",
    "for file in Files: \n",
    "    os.rename(file, 'arbolBLOSUM')\n",
    "\n",
    "Phyml_runDH = PhymlCommandline(input=phyfile_AA, datatype=\"aa\", model='Dayhoff')\n",
    "os.system(str(Phyml_runDH)) \n",
    "Files = glob.glob('*tree.txt') \n",
    "for file in Files: \n",
    "    os.rename(file, 'arbolDayhoff')\n",
    "\n",
    "##compute pairwise distances, store in dictionary\n",
    "\n",
    "K80_dict = distances(\"arbolK80\")\t\t\n",
    "HKY85_dict = distances(\"arbolHKY85\")\t\t\n",
    "JC69_dict = distances(\"arbolJC69\")\n",
    "GTR_dict = distances(\"arbolGTR\")\n",
    "BLOSUM_dict = distances(\"arbolBLOSUM\")\n",
    "DH_dict\t= distances(\"arbolDayhoff\")\n",
    "\n",
    "#use dictionary to loop through all results, print a result table that looks like\n",
    "#t1 t2 trans trans_of_total trvers tvers_of_total k80 hky85 jc69 gtr blosum dh deltamw deltapi deltacharge\n",
    "\n",
    "tr_of_total_list = []\n",
    "tv_of_total_list = []\n",
    "nucleotide_score_list = []\n",
    "BLOSUM62_score_list = []\n",
    "PAM60_score_list = []\n",
    "d_K80_list = []\n",
    "d_HKY85_list = []\n",
    "d_JC69_list = []\n",
    "d_GTR_list = []\n",
    "d_BLOSUM_list = []\n",
    "d_DH_list = []\n",
    "delta_mw_list = []\n",
    "delta_pI_list = []\n",
    "delta_charge_list = []                \n",
    "\n",
    "handle = open(\"results.txt\", \"w\")\n",
    "\n",
    "if len(sys.argv) == 3:\n",
    "    reference = int(sys.argv[2]) - 1\n",
    "    taxon1 = str(sorted_keys[reference])\n",
    "    for taxon2 in sorted_keys:\n",
    "        if str(taxon1) != str(taxon2):\n",
    "            #no if statement because this dictionary contains all pairs, even with itself, even redundant\n",
    "            tr_of_total = tr_tv_dict[taxon1][taxon2][1]       \n",
    "            tv_of_total = tr_tv_dict[taxon1][taxon2][3]\n",
    "            nucleotide_score = tr_tv_dict[taxon1][taxon2][6]\n",
    "            BLOSUM62_score = AA_score_dict[taxon1][taxon2][0]\n",
    "            PAM60_score = AA_score_dict[taxon1][taxon2][1]\n",
    "            d_K80 = extrapolate_distance(K80_dict, taxon1, taxon2)\n",
    "            d_HKY85 = extrapolate_distance(HKY85_dict, taxon1, taxon2)\n",
    "            d_JC69 = extrapolate_distance(JC69_dict, taxon1, taxon2)\n",
    "            d_GTR = extrapolate_distance(GTR_dict, taxon1, taxon2)\n",
    "            d_BLOSUM = extrapolate_distance(BLOSUM_dict, taxon1, taxon2)\n",
    "            d_DH = extrapolate_distance(DH_dict, taxon1, taxon2)\n",
    "            \n",
    "            #no if statement because this puts together stuff using single protein characteristics, hence no pairs in the dictionary         \n",
    "            delta_mw = round(float(PROT_info_dict[taxon1][0] - PROT_info_dict[taxon2][0]),4)\n",
    "            delta_pI = round(float(PROT_info_dict[taxon1][1] - PROT_info_dict[taxon2][1]),4)\n",
    "            delta_charge = round(float(PROT_info_dict[taxon1][2] - PROT_info_dict[taxon2][2]),4)\n",
    "            \n",
    "            tr_of_total_list.append(tr_of_total)\n",
    "            tv_of_total_list.append(tv_of_total)\n",
    "            nucleotide_score_list.append(nucleotide_score)\n",
    "            BLOSUM62_score_list.append(BLOSUM62_score)\n",
    "            PAM60_score_list.append(PAM60_score)\n",
    "            d_K80_list.append(d_K80)\n",
    "            d_HKY85_list.append(d_HKY85)\n",
    "            d_JC69_list.append(d_JC69)\n",
    "            d_GTR_list.append(d_GTR)\n",
    "            d_BLOSUM_list.append(d_BLOSUM)\n",
    "            d_DH_list.append(d_DH)\n",
    "            delta_mw_list.append(delta_mw)\n",
    "            delta_pI_list.append(delta_pI)\n",
    "            delta_charge_list.append(delta_charge)\n",
    "                            \n",
    "            print >> handle, taxon1, taxon2, tr_of_total, tv_of_total, nucleotide_score, BLOSUM62_score, PAM60_score, d_K80, d_HKY85, d_JC69, d_GTR, d_BLOSUM, d_DH, delta_mw, delta_pI, delta_charge\n",
    "    handle.close()\n",
    "            \n",
    "elif len(sys.argv) == 2: \n",
    "    for t1 in sorted_keys:\n",
    "        #print (\"taxon 1 is \" + str(t1))\n",
    "        for t2 in K80_dict[t1]:\n",
    "            taxon1 = t1\n",
    "            taxon2 = t2\n",
    "            #print (\"taxon 2 is \" + str(t2))\n",
    "            tr_of_total = tr_tv_dict[taxon1][taxon2][1]\n",
    "            tv_of_total = tr_tv_dict[taxon1][taxon2][3]\n",
    "            nucleotide_score = tr_tv_dict[taxon1][taxon2][6]\n",
    "            BLOSUM62_score = AA_score_dict[taxon1][taxon2][0]\n",
    "            PAM60_score = AA_score_dict[taxon1][taxon2][1]\n",
    "            d_K80 = extrapolate_distance(K80_dict, taxon1, taxon2)\n",
    "            d_HKY85 = extrapolate_distance(HKY85_dict, taxon1, taxon2)\n",
    "            d_JC69 = extrapolate_distance(JC69_dict, taxon1, taxon2)\n",
    "            d_GTR = extrapolate_distance(GTR_dict, taxon1, taxon2)\n",
    "            d_BLOSUM = extrapolate_distance(BLOSUM_dict, taxon1, taxon2)\n",
    "            d_DH = extrapolate_distance(DH_dict, taxon1, taxon2)\n",
    "                    \n",
    "            #no if statement because this puts together stuff using single protein characteristics, hence no pairs in the dictionary\n",
    "            delta_mw = round(float(PROT_info_dict[taxon1][0] - PROT_info_dict[taxon2][0]),4)\n",
    "            delta_pI = round(float(PROT_info_dict[taxon1][1] - PROT_info_dict[taxon2][1]),4)\n",
    "            delta_charge = round(float(PROT_info_dict[taxon1][2] - PROT_info_dict[taxon2][2]),4)\n",
    "                                                                                                                                                                                    \n",
    "            tr_of_total_list.append(tr_of_total)\n",
    "            tv_of_total_list.append(tv_of_total)\n",
    "            nucleotide_score_list.append(nucleotide_score)\n",
    "            BLOSUM62_score_list.append(BLOSUM62_score)\n",
    "            PAM60_score_list.append(PAM60_score)\n",
    "            d_K80_list.append(d_K80)\n",
    "            d_HKY85_list.append(d_HKY85)\n",
    "            d_JC69_list.append(d_JC69)\n",
    "            d_GTR_list.append(d_GTR)\n",
    "            d_BLOSUM_list.append(d_BLOSUM)\n",
    "            d_DH_list.append(d_DH)\n",
    "            delta_mw_list.append(delta_mw)\n",
    "            delta_pI_list.append(delta_pI)\n",
    "            delta_charge_list.append(delta_charge)\n",
    "\n",
    "            print >> handle, taxon1, taxon2, tr_of_total, tv_of_total, nucleotide_score, BLOSUM62_score, PAM60_score, d_K80, d_HKY85, d_JC69, d_GTR, d_BLOSUM, d_DH, delta_mw, delta_pI, delta_charge\n",
    "    handle.close()\n",
    "\n",
    "\n",
    "my_plot(\"distances_NUCvsDelta_MW.pdf\", delta_mw_list, d_K80_list, \"K80\", d_HKY85_list,\"HKY85\", d_JC69_list,\"JC69\", d_GTR_list, \"GTR\")\n",
    "my_plot(\"distances_NUCvsDelta_pI.pdf\", delta_pI_list, d_K80_list, \"K80\", d_HKY85_list,\"HKY85\", d_JC69_list,\"JC69\", d_GTR_list, \"GTR\")\n",
    "\n",
    "my_plot(\"distances_PROTvsDelta_MW.pdf\", delta_mw_list, d_BLOSUM_list, \"BLOSUM\", d_DH_list,\"Dayhoff\")\n",
    "my_plot(\"distances_PROTvsDelta_pI.pdf\", delta_pI_list, d_BLOSUM_list, \"BLOSUM\", d_DH_list,\"Dayhoff\")\n",
    "\n",
    "my_plot(\"NUCscorevsDelta_MW.pdf\", delta_mw_list, nucleotide_score_list, \"Nucleotide Score\")\n",
    "my_plot(\"NUCscorevsDelta_pI.pdf\", delta_pI_list, nucleotide_score_list, \"Nucleotide Score\")\n",
    "\n",
    "my_plot(\"PROTscorevsDelta_MW.pdf\", delta_mw_list, BLOSUM62_score_list, \"BLOSUM62\", PAM60_score_list,\"PAM60\")\n",
    "my_plot(\"PROTscorevsDelta_pI.pdf\", delta_pI_list, BLOSUM62_score_list, \"BLOSUM62\", PAM60_score_list,\"PAM60\") \n",
    "\n",
    "my_plot(\"transitions_of_totalvsDelta_MW.pdf\", delta_mw_list, tr_of_total_list, \"transitions\")\n",
    "my_plot(\"transitions_of_totalvsDelta_pI.pdf\", delta_pI_list, tr_of_total_list, \"transitions\")\n",
    "\n",
    "my_plot(\"transversions_of_totalvsDelta_MW.pdf\", delta_mw_list, tv_of_total_list, \"transversions\")\n",
    "my_plot(\"transversions_of_totalvsDelta_pI.pdf\", delta_pI_list, tv_of_total_list, \"transversions\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
